src/connectors/impls/clickhouse.rs

```diff
@@ -0,0 +1,290 @@
+
+mod conversion;
+
+use std::fmt::{self, Display, Formatter};
+
+use crate::connectors::{prelude::*, utils::url::ClickHouseDefaults};
+
+use clickhouse_rs::{
+    types::{DateTimeType, SqlType},
+    Block, Pool,
+};
+use either::Either;
+
+#[derive(Default, Debug)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "clickhouse".into()
+    }
+
+    async fn build(&self, alias: &str, raw_config: &ConnectorConfig) -> Result<Box<dyn Connector>> {
+        let config = raw_config
+            .config
+            .as_ref()
+            .ok_or_else(|| ErrorKind::MissingConfiguration(alias.to_string()))?;
+
+        let config = ClickhouseConfig::new(config)?;
+
+        Ok(Box::new(Clickhouse { config }))
+    }
+}
+
+pub struct Clickhouse {
+    config: ClickhouseConfig,
+}
+
+#[async_trait::async_trait]
+impl Connector for Clickhouse {
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let db_url = self.connection_url();
+        let columns = self
+            .config
+            .columns
+            .iter()
+            .map(|Column { name, type_ }| (name.clone(), type_.clone()))
+            .collect();
+
+        let sink = ClickhouseSink {
+            db_url,
+            pool: None,
+            columns,
+        };
+        builder.spawn(sink, sink_context).map(Some)
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Structured
+    }
+}
+
+impl Clickhouse {
+    fn connection_url(&self) -> String {
+        let url = self.config.url.as_str();
+        let compression = self.config.compression.unwrap_or_default();
+
+        format!("{url}?compression={compression}")
+    }
+}
+
+#[derive(Deserialize)]
+struct ClickhouseConfig {
+    url: Url<ClickHouseDefaults>,
+    compression: Option<Compression>,
+    columns: Vec<Column>,
+}
+
+impl ConfigImpl for ClickhouseConfig {}
+
+#[derive(Clone, Copy, Deserialize)]
+#[serde(rename_all = "snake_case")]
+enum Compression {
+    None,
+    Lz4,
+}
+
+impl Default for Compression {
+    fn default() -> Compression {
+        Compression::None
+    }
+}
+
+impl Display for Compression {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            Compression::None => "none",
+            Compression::Lz4 => "lz4",
+        }
+        .fmt(f)
+    }
+}
+
+#[derive(Deserialize)]
+struct Column {
+    name: String,
+    #[serde(rename = "type")]
+    type_: DummySqlType,
+}
+
+pub(crate) struct ClickhouseSink {
+    db_url: String,
+    pool: Option<Pool>,
+    columns: Vec<(String, DummySqlType)>,
+}
+
+#[async_trait::async_trait]
+impl Sink for ClickhouseSink {
+    async fn connect(&mut self, _ctx: &SinkContext, _attempt: &Attempt) -> Result<bool> {
+        let pool = Pool::new(self.db_url.as_str());
```

Comment:
Is there a way we can verify here in `connect` that the connection was successfully established?

Reply:
One solution would be to store a [`ClientHandle`] instead of storing a [`Pool`] in the `ClickhouseSink` structure. I think that's the best way to proceed.

[`ClientHandle`]: https://docs.rs/clickhouse-rs/1.0.0-alpha.1/clickhouse_rs/struct.ClientHandle.html
[`Pool`]: https://docs.rs/clickhouse-rs/1.0.0-alpha.1/clickhouse_rs/pool/struct.Pool.html

Reply:
This should be fixed by [d948725](https://github.com/tremor-rs/tremor-runtime/pull/1538/commits/d9487251771ab07d16b4caac11e9789ef8616ce0). Feel free to reopen if needed :)